### ISSUE TITLE
add QUIC support

### ISSUE THREAD

**Issue by joeblew99** [nats-io/gnatsd#457]
https://github.com/lucas-clemente/quic-go
quic is a network protocol using UDP.
fast and NAT aware. So great for Mobile.
it also works in Chrome browsers. Still might need flags
It also has been submitted to the IETF

**Comment by lenovouser** [comment]
I am pretty sure NATS is using their own protocol? What is QUIC going to change there?

**Comment by mcqueary** [comment]
*TL;DR It's not clear that the benefits of QUIC vs TCP are clear-cut for a connection-oriented protocol like NATS.*

The primary value of QUIC (as I understand it) is in reducing latency for secure, browser-style (i.e. connectionless) interactions with a web server, where every single interaction involves connecting, requesting data and getting a response. To this end, one of its key advantages is in multiplexing streams over a single connection for efficiency, thereby sharply reducing the kind of per-stream connection overhead typically associated with (e.g.) a connectionless protocol like HTTP, and improving compression performance.

So here's the rub: NATS is a connection-oriented protocol.  It needs but a single TCP connection to service multiple streams data, regardless of whether it's request/reply or streaming data in one direction or the other. The overhead that QUIC eliminates is chiefly in reducing round trip times (RTT) associated with setting up a connection and sending a request-reply _every single time_ you interact with the server. While you _could_ use NATS that way, and I'm sure some people do, I think the majority of users employ a small number of long-lived connections for most of their needs, vs. creating new connections for every new request.

**Comment by joeblew99** [comment]
@mcqueary 
Thanks for the due diligence on QUIC in relation to NATS.

i ended up writing my own semantic addressing layer above NATS.. But here is why i did it.

I raised it because i wanted to use NATS behind a proxy (any golang one) with mobile and IOT client devices on mobile telephone networks. 
QUIC is UDP based and thats a big bonus on Mobile networks.
The two types of flows are:
- client sending in sensor data over QUIC that is stored in minio before processing.
- clients participating in shared data that is changing often. When one client changes common data, it sends the mutation to NATS and NATS rebroadcasts it to all client subscribers. 

I also use loRaWAN for IOT and SMPP for mobile devices. QUIC cant help me there because these networks cant use QUIC so  use a different proxy in front of NATS. 
So i tend to use different gateways / proxies in front of NATS depending on the protocol, with NATS being the core message gateway, and then logic behind that.

So, what i was really asking for is semantic addressing in essence.

**Comment by joeblew99** [comment]
I closed this because I ended up just making wrappers for QUIC and other network transports and bindings.

Used go-kit which is very good for these situations with some go generators.
The generators run of protocol buffers with annotations hinting to the code generator writers what topics to use on Nats.

In hindsight this would be much easier if NATS supported running embedded properly because there would be less network layers and so would be much higher performance.
But I could not work out a way to modify Nats to use go routines instead of TCP. So I was for ed to run Nats on a sperate process from the gateway.

@ mcqueary did I explain this well enough ?
Basically I don't want you to change what training sports bars uses, I just would like to be able to run Nats embedded and allow the pub sub to use go routines, instead of tcpip.

**Comment by ColinSullivan1** [comment]
@joeblew99 , I can see the advantage of a pluggable protocol, but this won't in the roadmap of NATS at this time.  The beauty of OSS is that you can fork to develop on your own;  if you implement something along these lines, we'd certainly be interested in checking it out.

**Comment by jdrews** [comment]
I noticed that the roadmap for NATS lists QUIC Support in 2021-Q3. Very excited about this! 
![image](https://user-images.githubusercontent.com/172766/107681433-09685680-6c6d-11eb-854a-6a774d69a685.png)
https://nats.io/about/

Can someone from the NATS team point me to more information about the plan for QUIC support?

**Comment by jdrews** [comment]
I jumped on the NATS Slack and talked to the NATS maintainers. No details yet. Sounded like the goal was to make QUIC Support transparent to client and servers.

**Comment by jdrews** [comment]
Looks like QUIC has now moved on the roadmap to 2021-Q4
https://nats.io/about/

**Comment by KristoferHansson** [comment]
@derekcollison: Any updated info on QUIC support, still planned for Q3?
We are working with indoor fleets of robots with shared wifi to backend, coverage and load varies so QUIC is of high interest.
Currently looking at the recently added QUIC support in Zenoh: https://zenoh.io/docs/manual/quic/

**Comment by KristoferHansson** [comment]
Very much looking forward to the QUIC + Advanced Edge / 5G Support previously planned for Q3
We are operating fleet of robots indoors with shared WiFi link to backend, highly varying radio conditions, why QUIC is of high interest. Currently running reliable UDP implementation to backend(where we use NATS).

**Comment by derekcollison** [comment]
I am interested in learning more about your situation. Since RFC-9000 just made it out we would imagine basing on something around that. It will take some time for the ecosystem to get everything updated etc.

Is your interest around QUIC about noisy comm paths or head of line blocking?

**Comment by jdrews** [comment]
@derekcollison My use case for QUIC on NATS is definitely the "noisy comm paths". In particular I'm interested in using NATS JetStream with QUIC for SATCOM between datacenters and distant, ephemeral edge nodes that have low bandwidth (2mbps or less), high latency (650ms-2sec), high packet loss (5% or more), and intermittent access (weeks of downtime with queued up data).

You can read more about the use cases here:
https://tools.ietf.org/id/draft-kuhn-quic-4-sat-06.html
https://datatracker.ietf.org/meeting/103/materials/slides-103-maprg-quic-and-satcom-nicolas-kuhn-01

**Comment by derekcollison** [comment]
Will take a look for sure. We may be looking to reach out to partners and customers on this work as well to make sure we have a good understanding of what is desired.

If you have any interest shoot me an email.

**Comment by tysonrm** [comment]
Looks like support is on the roadmap. The advantage of QUIC, the transport protocol for HTTP/3, is that it does not block other streams in the event of a missing or malformed packet.

**Comment by xorgy** [comment]
> Looks like support is on the roadmap. The advantage of QUIC, the transport protocol for HTTP/3, is that it does not block other streams in the event of a missing or malformed packet.

Thanks, I thought that would be pretty obvious, but apparently it bears repeating.

**Comment by jdrews** [comment]
I don't see QUIC on the roadmap for 2023 anymore. However there is `Leaf Node Enhancements` in 2023-1H which I think might be QUIC. Is that the case? 
![image](https://user-images.githubusercontent.com/172766/200949063-26abaaa6-a6fc-42d8-8f2c-66a4950b93db.png)

**Comment by derekcollison** [comment]
Yes we are considering QUIC in that timeframe and if we proceed will start with leafnodes.

**Comment by hwinkel** [comment]
@derekcollison is the roadmap above still valid?

**Comment by derekcollison** [comment]
Needs updating.

**Comment by NoahHahm** [comment]
CoreNATS is TCP-based.
I wonder if the latency will be faster if QUIC-based technology is also available.
(Especially in super multi-cluster environment)
I'm curious if you've tried an experiment.

**Comment by derekcollison** [comment]
TCP with short RTT and no loss will beat QUIC. QUIC will have an advantage over long RTT and lossy networks.

**Comment by chrisprobst** [comment]
Some random details about QUIC, for those who don't know. QUIC is not faster than TCP and is has much more CPU overhead if used naively.

For curious reader:
- QUIC is UDP-based and high-throughput UDP (believe it or measure) is extremely CPU intensive because of a lot of syscalls (one per packet). The quic-go impl. is using a lot of [dirty tricks](https://quic-go.net/docs/quic/optimizations/) to make this less impactful but it is still nowhere as efficient as TCP. Try a simple loopback example with QUIC vs TCP. quic-go will peak out at 500-1000 MB/s, even with all tricks enabled and 100% CPU. TCP can easily deliver 7-10 GB/s on a loopback device with large buffers. And no, it's not a Go issue. It's the nature of UDP, it's expensive for high throughput. You can circumvent this by going even further and using kernel-bypass UDP which requires more than your usual Ubuntu ;-). So, long story short, QUIC is NOT here to beat TCP in scenarios where TCP is already king. It is meant to improve where TCP has some weaknesses like NAT traversal, handshake, head of line blocking, etc. For simple singular streams where bandwidth is of most importance, nothing will beat good old TCP.

Still, I believe QUIC (even raw UDP tbh.) could have a valid use case for NATS. Pub/Sub is already UDP-like in the sense that it is fire and forget. I work on systems where clients receive regular updates which always contains the latest snapshot, lost data is not an issue because every message will contain the latest state (basically games or media apps with regular key frames). NATS cannot be applied here because you don't want nor need the overhead of TCP. You want to pub and sub as fast as possible, lost messages are fine. NATS could support this usecase with such a protocol (QUIC in unreliable mode), even Browsers as end devices. It could open a whole new set of scenarios, like media and real-time media which is totally undoable so far.

Just a thought.

Chris

**Comment by remram44** [comment]
Why not use `sendmmsg()` instead of relying on segmentation offload?

**Comment by chrisprobst** [comment]
It's a good start but not the full answer. There is more to it. There are some very informative issues on the quic-go repo about this, or some details can also be found here: https://tailscale.com/blog/more-throughput. The wireguard-go project is a true source of wisdom for these things. But there is a ceiling, TCP is just unbeatable in this specific domain (CPU overhead vs. bandwidth on average OS configuration).

In general: You need a mixture of tricks to make it faster but you will never come close to TCP (in terms of CPU utilization and Kernel overhead) without full kernel bypass.

EDIT: The bottom line is: Don't use QUIC because you think it's faster (it's not). Use it for the right reasons and there are some very good reasons.

**Comment by xorgy** [comment]
@chrisprobst It really depends on the typical message sizes, the specific way that NATS is mapped to QUIC (stream per rpc? stream per subscription? stream per message?), and other things. 
As for syscall overhead, io_uring addresses that quite thoroughly. Zero copy send/receive on io_uring is... pretty fast, especially when there isn't much segmentation to do anyhow (a lot of messages fit in one packet, even with the added framing).
On the BSDs sure, you don't really have anything as efficient as you do on Linux yet... and if for some reason you are running this on Windows, IOCP may get you somewhere (not sure how it compares, that's not my expertise).

P.S. WireGuard is not necessarily the right comparison for this, as WireGuard's userspace issues relate a lot to transferring large packets (lots of segmentation) to relatively few peers, and the primary outcome they were measuring in that tailscale writeup was bulk throughput..

**Comment by chrisprobst** [comment]
io_uring and Go does not seem like a good choice. The Go Team Looked at it and won‘t add it anytime soon. Doing it yourself means bypassing the scheduler which is not ideal either. I will do a small Test with io_uring and UDP but i am very sure it does not solve the Problem I described. I am of Course Talking about high bandwidth scenarios. When you Are far away from congesting the net interface, I might agree. For high bandwidth scenarios, however, I cannot See how io_uring will help. UDP pacing is another Problem
That cannot be solved efficiently in User space.

**Comment by floatdrop** [comment]
> It is meant to improve where TCP has some weaknesses like NAT traversal, handshake, **head of line blocking**, etc. For simple singular streams where bandwidth is of most importance, nothing will beat good old TCP.

I want to double that point, especially about HOLB. In my use case I want to transfer UDP RTP packets through NATS and wrapping them in TCP connection strongly affects subscribers receive performance on not-so-well working cross-dc connections.

It is not about raw performance, but avoiding some TCP weaknesses.
